ROFO-205 searchFoodSpots 응답 데이터 추가

Unagi-zoso opened the pull request:
### 개요
기존 응답 데이터에 영업 시간, 썸네일 이미지, 푸드트럭 여부, 평균 평점, 리뷰 개수 추가 요청이 들어와 처리합니다.

### 변경사항
- 필요 쿼리 메소드 작성
- 기존 response 클래스 프로퍼티 추가
- 기존 로직 변경


### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-205) 


### 리뷰어에게 하고 싶은 말
썸네일의 겨우 한 장만 노출될 것 같아 findOne.. 하는 쿼리 메소드를 작성했습니다.
kotlin jdsl 에서 findOne 메소드 자체를 제공하는건 꺼려하여 [관련 PR](https://github.com/line/kotlin-jdsl/pull/531)
findAll 에 pageable 을 담아 하나만 반환 받도록 했습니다. [관련 이슈](https://github.com/line/kotlin-jdsl/issues/695)
(findAll 대신 getPage 사용 시 count 쿼리가 발생합니다.)

**영업시간의 부재** 시 금일 무슨 요일인지와 **영업시간은 00:00** 인 값을 반환합니다.
**리뷰의 부재** 시엔 리뷰 수 0개와 **평균 평점 0.0** 을 반환합니다.

모든 피드백 감사히 잘 받겠습니다.

Unagi-zoso committed:
ROFO-205 findOneByFoodSpots 메소드 및 테스트 작성

Unagi-zoso committed:
ROFO-205 findByFoodSpots 메소드 및 테스트 작성

Unagi-zoso committed:
ROFO-205 FoodSpotsFixture 수정

Unagi-zoso committed:
ROFO-205 searchFoodSpots 응답 데이터 추가

Unagi-zoso committed:
ROFO-205 테스트 스타일 수정 및 불필요한 stubbing 제거

shouldHaveSize 를 사용하는 편이 읽고 좋아 변경했습니다.

lin-chae commented on src/main/kotlin/kr/weit/roadyfoody/foodSpots/application/service/FoodSpotsQueryService.kt:
```diff
@@ -195,4 +214,25 @@ class FoodSpotsQueryService(
+            foodSpotsId = foodSpot.id,
+            dayOfWeek = dayOfWeek,
+            openingHours = "00:00",
+            closingHours = "00:00",
```
금일 영업시간 로직 잘 짜주셨네요! 👍

lin-chae commented on src/main/kotlin/kr/weit/roadyfoody/foodSpots/repository/FoodSpotsPhotoRepository.kt:
```diff
@@ -1,13 +1,39 @@
+) : CustomFoodSpotsPhotoRepository {
+    override fun findOneByFoodSpots(foodSpotsId: Long): FoodSpotsPhoto? =
+        kotlinJdslJpqlExecutor
+            .findAll(Pageable.ofSize(1)) {
```
findOne 메소드 자체를 제공하는건 꺼려하는군요. 알아갑니다 🤔

hyunjungkimm replied:
findOne 혹은 findFirst 같은 메소드를 제공하지 않는다는 사실 알아갑니다! 👍

konempty replied:
확장함수로 selectOne과 같은 함수를 만들어 두는게 좋을것 같아요!
그리고 요렇게 하면 불필요하게 count쿼리가 한번 날라가는걸로 알고 있는데 pagination이 아닌 방식으로 하나만 들고올순 없었을까요?

lin-chae reviewed (approved):
항상 일을 빨리 처리하시네요! 굳굳! 👍  고생하셨습니다!

hyunjungkimm reviewed (approved):
빠른 작업 👍 👍  고생하셨습니다!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/foodSpots/application/service/FoodSpotsQueryService.kt:
```diff
@@ -54,7 +55,10 @@ class FoodSpotsQueryService(
-    fun searchFoodSpots(foodSpotsSearchQuery: FoodSpotsSearchCondition): FoodSpotsSearchResponses {
+    fun searchFoodSpots(
+        foodSpotsSearchQuery: FoodSpotsSearchCondition,
+        today: LocalDate = LocalDate.now(),
```
이름부터 today이고 기본값도 오늘날짜 밖에 안쓰는것 같은데 
굳이 메소드 파라메터로 받을 필요가 있을까??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/foodSpots/application/service/FoodSpotsQueryService.kt:
```diff
@@ -67,13 +71,28 @@ class FoodSpotsQueryService(
         val foodSpotsSearchResponses =
             result.map { foodSpots ->
                 val openValue: OperationStatus = determineOpenStatus(foodSpots)
+                val foodSpotsPhoto = foodSpotsPhotoRepository.findOneByFoodSpots(foodSpots.id)
+                val reviews = reviewRepository.findByFoodSpots(foodSpots)
+                val averageRating =
+                    if (reviews.isNotEmpty()) {
+                        val sum = reviews.sumOf { it.rate }.toDouble()
+                        ((sum / reviews.size) * 10).toInt() / 10.0
+                    } else {
+                        0.0
+                    }
+
```
요건 DB에서 계산해서 가저오는게 어떨까요??
아마 FoodSpots를 온전히 쓰고 있어서 그런것 같은데 프로젝션으로 가져오는게 좋을것 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/foodSpots/application/service/FoodSpotsQueryService.kt:
```diff
@@ -195,4 +214,25 @@ class FoodSpotsQueryService(
+
+    private fun getTodayOperationHoursResponse(
+        foodSpot: FoodSpots,
+        today: LocalDate = LocalDate.now(),
```
요기도 today를 파라메터로 받는 이유가 궁금합니다~!

Unagi-zoso replied:
해당 기능이 LocalDate 를 의존하는데 이걸 테스트하려면 외부로부터 주입 받는게 좋아보이더라구요.

테스트 내에서는 LocalDate.get() 을 고정된 값을 반환하도록 stubbing 하였는데 

모킹 없이 테스트 대상(foodSpots) 의 소유 운영시간을 오늘 요일에 상대적으로 구성해 테스트를 구성할까 싶다가도 자정이 넘어 요일이 바뀐다같은 상황에 대한 대처가 어렵더라구요. 자칫하면 어쩔땐 성공하고 어쩔땐 실패하는 테스트가 될 수 있을것 같았습니다.

그래서 지금처럼 모킹을 해서 보여주는 편이 직관적이고 결과를 쉽게 예상할 수 있지 않을까 생각했습니다.  

이게 `seachFoodSpots` 의 패러미터까지 LocalDate 를 만들어둬서 **테스트 코드만을 위한
비즈니스 코드의 변경**이 아닌가 우려되어 `getTodayOperationHoursResponse` 를 **따로 클래스**로 나누던가 **public** 으로 만들던가 하는식으로해서 `getTodayOperationHoursResponse` 의 패러미터에만 LocalDate 를 주입 받을 수 있게해 테스트를 할까 싶다가 따로 **성급하게 클래스를 만들기**도 그렇고 **private 하게 사용되는 메소드를 테스트를 위해 public** 으로 바꾸는 것도 마음에 걸렸습니다.

konempty replied:
그냥 요일을 다내려주고 클라이언트가 알아서 날짜별로 쓰라고 하면 서버에서 오늘의 요일을 알아낼 필요가 없지않을까요??

Unagi-zoso replied:
클라이언트분이 응답 필드 추가할 때 운영시간 중 오늘 것만을 요구하셔서
오늘 것만 내릴 수 있게 만들어봤습니다. 

현재 시간에 의존하는 로직은 서버에서 다루지 않는 편이 좋을까요?

(응답 필드 추가사항)
<img width="161" alt="스크린샷 2024-09-07 오후 4 41 39" src="https://github.com/user-attachments/assets/367feb2a-65c1-4e22-8dd4-7eeb29659424">

konempty replied:
테스트를 위한 코드가 프로덕션에 영향을 끼치는것도 싫고
클라가 할수 있는 일인것 같아서 그랬습니다!

konempty replied:
아 생각해보니까 mockkStatic이라는 메소드를 쓰면 스태틱메소드도 모킹이 가능한데 그걸로도 안되나요??

konempty replied:
참고 : https://deep-dive-dev.tistory.com/84

Unagi-zoso replied:
이런 환상의 기능이 존재하다니 한 번 시도해볼게요. 감사합니다!

konempty reviewed (commented):
고생하셨습니다 👍 👍 
궁금한게 몇가지 있어서 코멘트 남깁니다~!

Unagi-zoso commented:
> 그리고 요렇게 하면 불필요하게 count쿼리가 한번 날라가는걸로 알고 있는데 pagination이 아닌 방식으로 하나만 들고올순 없었을까요?

@konempty 
제가 알고있는 바론 **findAll** 에 대한 pageable 요청의 경우 count 쿼리가 호출되지 않는 것으로 알고 있습니다.
내부 원리는 아직 잘 모르겠습니다만 제가 확인했을 때도 추가적인 count 쿼리는 날라가지 않았습니다.
**getPage** 에 대해 pageable 요청을 했을땐 count 쿼리가 날라갔습니다.

참고한 자료에 있던 부분을 인용했습니다. [출처 (Kotlin Jdsl issue) 맨 하단](https://github.com/line/kotlin-jdsl/issues/695)
```
setMaxResults를 사용하고 싶지 않으시다면 findAll(pageable)을 사용하시는 방법도 있습니다. findAll의 경우는 count 쿼리를 호출하지 않습니다.
```

혹시 따로 추천하시는 방법이 있으실까요?

konempty commented:
@Unagi-zoso 
ㅋㅋㅋㅋ 의도는 이해가 되는데 JDSL 문법이 오해하기 딱 좋아보이네요
https://github.com/line/kotlin-jdsl/pull/764
그래서 그런가 이번 업데이트에서 limit 파라메터가 추가됐군요

Unagi-zoso committed:
ROFO-205 findOne 확장함수 생성 및 적용

Unagi-zoso committed:
ROFO-205 피드백 반영

기존에 Review 수, 평균평점을 얻어오는 쿼리 메소드가 있어 이를 사용하기로 했습니다. (getReviewAggregatedInfo)

LocalDate 를 외부에서 주입 받지 않고 Static 메소드를 모킹하기로 했습니다.

static mock 객체 생성 시 이름없는 모듈에서도 리플렉션 접근할 수 있게 gradle test task 에 추가합니다.

Unagi-zoso committed:
ROFO-205 사용되지 않는 메소드 및 테스트 제거

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/utils/KotlinJDSLUtil.kt:
```diff
@@ -32,3 +32,10 @@ fun <T : Any> KotlinJdslJpqlExecutor.getPage(
+
+private val SINGLE_RESULT_PAGEABLE = Pageable.ofSize(1)
+
+// 3.5.2 버전부턴 findAll 의 parameter 에 limit 이 추가됩니다.
+// 참고 : https://github.com/line/kotlin-jdsl/releases/tag/3.5.2
```
👍

konempty reviewed (approved):
고생하셨습니다 👍 👍

Unagi-zoso committed:
Merge branch 'develop' into feature/ROFO-205
